classes/session_manager.php

```diff
@@ -0,0 +1,261 @@
+            'observer' => $sessiondata->observer_id
+        ];
+    }
+
+    /**
+     * Returns the observation point data for a session, including any existing responses to the points
+     * @param int $sessionid observation session id
+     * @return array array of observation points
+     */
+    public static function get_session_data(int $sessionid) {
+        // Get the details for the session.
+        $sessioninfo = self::get_session_info($sessionid);
+        $obid = $sessioninfo['obid'];
+
+        // Get all points.
+        $observationpoints = \mod_observation\observation_manager::get_points_and_responses($obid, $sessionid);
+        return $observationpoints;
+    }
+
+    /**
+     * Returns a list of incomplete points. I.e. those without a response.
+     * @param int $sessionid observation sesssion ID
+     * @return array array containing points with no responses.
+     */
+    public static function get_incomplete_points(int $sessionid) {
+        $sessiondata = self::get_session_data($sessionid);
+        $incompletepoints = array_filter($sessiondata, function($point) {
+            return $point->response_id === null;
+        });
+        return $incompletepoints;
+    }
+
+    /**
+     * Calculates the total grade given for an observation session.
+     * @param int $sessionid observation session id
+     * @return array array containing two values: total and max. Total is the total grade given, Max is the total grade available.
+     */
+    public static function calculate_grade(int $sessionid) {
+        $sessioninfo = self::get_session_info($sessionid);
+        $obid = $sessioninfo['obid'];
+        $observationpoints = \mod_observation\observation_manager::get_points_and_responses($obid, $sessionid);
+
+        $maxgrades = array_column($observationpoints, 'max_grade');
+        $givengrades = array_column($observationpoints, 'grade_given');
+
+        // Ensure for each grade given that it is less than the max_grade.
+        foreach ($givengrades as $index => $grade) {
+            if ($grade > $maxgrades[$index]) {
+                throw new \moodle_exception("One or more grades was greater than the max grade.
+                    This can happen when an observation point is editied before a session has been submitted.");
+            }
+        }
+
+        $totalmaxgrade = array_sum($maxgrades);
+        $totalgradegiven = array_sum($givengrades);
+
+        return [
+            'total' => $totalgradegiven,
+            'max' => $totalmaxgrade
+        ];
+    }
+
+    /**
+     * Saves extra comment for an observation sesssion
+     * @param int $sessionid observation session id
+     * @param string $extracomment extra comment
+     */
+    public static function save_extra_comment(int $sessionid, string $extracomment) {
+        global $DB;
+        $DB->update_record('observation_sessions', ['id' => $sessionid, 'ex_comment' => $extracomment]);
+        return;
```

Comment:
the return here is unnecessary, but perhaps it should return the true or false from the update?